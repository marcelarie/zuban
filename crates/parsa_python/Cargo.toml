[package]
name = "parsa_python"
version.workspace = true
edition.workspace = true
rust-version.workspace = true
publish = false
homepage.workspace = true
authors.workspace = true

[dependencies]
parsa.workspace = true
lazy_static.workspace = true
regex.workspace = true

[dev-dependencies]
insta.workspace = true
test_utils.workspace = true
