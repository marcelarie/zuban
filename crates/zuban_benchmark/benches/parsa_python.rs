use criterion::{criterion_group, criterion_main, Criterion};

use parsa_python_cst::Tree;

fn parse_file(copies: usize) {
    let some_code = utils::dedent(
        r#"
        import os
        def x(a, b, c):
            try:
                return a
            except Exception:
                return [1, 2, 3, 4, 5, 6]
        class C:
            x: int
            y = 1 # type: int
            z: dict[int, list[str]] = {1: ["", r''], 2: [f""" {1} """]}
            def f(self, x: list[str], y: int): ...
    "#,
    );
    assert_eq!(some_code.len(), 262);
    Tree::parse(some_code.repeat(copies).into());
}

fn bench_parser(c: &mut Criterion) {
    c.bench_function("parse big file", |b| b.iter(|| parse_file(1000)));
    c.bench_function("parse small file", |b| b.iter(|| parse_file(1)));
}

// Register the benchmarks
criterion_group!(benches, bench_parser);
criterion_main!(benches);
