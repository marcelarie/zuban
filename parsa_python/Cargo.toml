[package]
name = "parsa_python"
version = "0.1.0"
edition.workspace = true
rust-version.workspace = true
homepage.workspace = true
authors.workspace = true

[dependencies]
parsa.workspace = true
lazy_static.workspace = true
regex.workspace = true

[dev-dependencies]
insta.workspace = true
