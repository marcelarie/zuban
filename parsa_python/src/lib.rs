use parsa::{create_parser,create_node,create_token};


create_token!(PythonToken, enum TokenType, String, Number, Endmarker);
create_node!(PythonNode, enum NodeType, file_input, foo);

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn it_works() {
        assert_eq!(2 + 2, 4);
        parse_python("bar");
        dbg!(TokenType::get_map());
        dbg!(NodeType::get_map());
        return
    }
}

struct PythonTokenizer {
}

impl PythonTokenizer {
    fn new_tok(&self) -> PythonToken {
        PythonToken {
            start: 0,
            length: 0,
            type_: TokenType::Endmarker,
        }
    }
}

impl parsa::Tokenizer<PythonToken> for PythonTokenizer {
    fn new(_code: &str) -> Self {
        Self {}
    }

    fn yield_next(&self) -> PythonToken {
        self.new_tok()
    }

}

create_parser!(parse_python, PythonTree, PythonNode,
               PythonToken, PythonTokenizer, TokenType, NodeType);
//create_parser!(parse_pythonx, PythonTreex, PythonNodex,
//               PythonToken, PythonTokenizer, TokenType, NodeType);
