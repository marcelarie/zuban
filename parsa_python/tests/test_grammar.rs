extern crate insta;

use parsa_python::*;

fn dedent(s: &'static str) -> String {
    let mut indent = usize::MAX;
    let lines: &Vec<_> = &s.split('\n').collect();
    for line in lines {
        if !line.trim().is_empty() {
            indent = std::cmp::min(indent, line.len() - line.trim_start().len());
        }
    }
    if indent == usize::MAX {
        return s.to_string();
    }
    let new_lines: Vec<_> = lines
        .iter()
        .map(|line| {
            if line.len() >= indent {
                &line[indent..]
            } else {
                line
            }
        })
        .collect();
    new_lines.join("\n")
}

fn tree_to_string(tree: PyTree) -> String {
    fn recurse(code: &mut String, node: &PyNode, depth: usize) {
        *code += &" ".repeat(depth);
        *code += &format!(
            "{}: {}-{}{}\n",
            node.type_str(),
            node.start(),
            node.end(),
            if node.is_leaf() {
                format!(" {:?}", node.as_code())
            } else {
                "".to_string()
            }
        );
        for c in node.iter_children() {
            assert_eq!(node.index, c.parent().unwrap().index);
            recurse(code, &c, depth + 1);
        }
    }

    let root = tree.root_node();
    assert!(root.parent().is_none());
    let mut code = String::new();
    recurse(&mut code, &root, 0);
    code
}

macro_rules! parametrize_snapshots {
    ($($name:ident : $input:expr;)*) => {$(
        #[test]
        fn $name() {
            let tree = parse($input.into());
            insta::assert_snapshot!(stringify!($name), tree_to_string(tree));
        }
    )*}
}

parametrize_snapshots!(
    empty: "".to_owned();
    very_simple: "1".to_owned();
    json: "{foo: 1}\n".to_owned();
    simple: dedent("
        a(bar:=1, foo=1);
        if a:
         1
        else: 2
        ");
    assignments_simple: dedent("
        a = 1
        a, b = 2
        a, b, = 2
        (a, b) = 2
        ((a, b)) = 2
        a = b = 3
        a = (b, c) = 3
        a = (b, c,) = 3
        a = b, c, = 3
        ");
    assignments_star: dedent("
        *foo = *bar = *baz
        i, *foo = ((j, [*bar, k])) = *baz
        *foo, *bar = yield baz
        ");
    assignments_aug: dedent("
        foo += 3
        foo.bar |= 3
        foo, bar += 3
        ");
    assignments_setitem: dedent("
        foo[0] += 3
        foo[0] = a = 3
        foo[a, b] = c
        foo[a::, b:f, c:=3] = c
        ");
    assignments_annotation: dedent("
        foo: bar
        foo: int = 3
        foo.bar: baz
        foo: bar = a = 4
        ");
    assignments_with_call: dedent("
        foo().bar = 1
        foo(1,).bar[1] = 1
        foo(a for a in b).bar[1] = 1
        ");
    for_stmt: dedent("
        for x in [1,2]:
            ...
        for *b in *c: ...
        for c.x, (*b, d.y[0]) in i, *j, k: ...
        ");
    comprehension_simple: dedent("
        [1 for a in b]
        [a for *a, c in b if c if d]
        foo(a for (*a, (*c, b),), in d)
        ");
    comprehension_fails: dedent("
        # fails
        [1 for a in *b]
        [1 for a in b if a else c]
        ");
    del_stmt: dedent("
        del foo
        del foo,
        del foo, bar.baz
        del foo[1]
        ");
    cls: dedent("
        class Foo(object):
            def __init__(self, /, f, *, g):
                pass

            @property
            def prop(self):
                pass
        ");
    import_names: dedent("
        import foo
        import foo.bar as baz
        import foo, bar, baz
        import foo.bar.baz
        ");
    import_name_fails: dedent("
        import foo,
        import foo, bar,
        ");
    import_from_normal: dedent("
        from foo import bar
        from foo import bar as baz
        from foo.bar.baz import *
        ");
    import_from_dotted: dedent("
        from . import (foo as baz, bar)
        from . import (foo, bar,)
        from ..foo import bar as baz
        from ... import *
        ");
    import_from_fails: dedent("
        from . import foo,
        from . import foo, bar,
        ");
    calls: dedent("
        foo(bar)
        foo(bar_foo=3)
        foo(bar_baz:=2)
        foo(**kwargs)
        ");
    nested_call: dedent("
        foo(1)()
        ");
    complex_calls: dedent("
        foo(bar, 3)
        foo(bar, *foo)
        foo(bar, *foo, name=3)
        foo(bar, *foo, name=3, **kwargs)
        foo(bar, *foo, name=3, **kwargs, b=3)
        foo(**kwargs, b=3, **foo)
        ");
    operators: dedent("
        f | d | c + 5 & 2 * 3 * 4 ** 5 ** 6
        ");
    simple_error_recovery: dedent("
        a + 3 +
        b = 3
        ");
    terminal_error_recovery: dedent("
        ?
        ");
    terminal_and_nonterminal_error_recovery: dedent("
        foo
        (1) + ? + c
        bar
        ");
    keyword_recovery: dedent("
        foo
        def
        bar
        else:
        baz
        ");
    match_simple: dedent("
        match foo:
            case ['a']: ...
            case _:
                pass
        ");
    soft_keyword_underscore: dedent("
        match _:
            case ['a']: _(_)
            case _:
                _
        match = 3
        ");
    match_underscore_with_as: dedent("
        match foo:
            case bar as _: pass
        ");
    match_class_pattern: dedent("
        match a:
            case A():
                pass
            case B(1, 2):
                pass
            case B(1, b=2):
                pass
            case B(a=1, b=2):
                pass
        ");
    dict_literal1: dedent("
        {1: 2}
        {**foo}
        {**foo, 1: 2}
        {**foo, **bar}
        ");
    function_simple: dedent("
        def foo(): ...
        def foo(a: 3): ...
        def foo(a, b,): ...
        def foo(a: f = 3, b=4): ...
        def foo(a=3, b=4): ...
        ");
    function_args: dedent("
        def foo(a, *, b=3): ...
        def foo(a=3, *, c): ...
        def foo(a, *, x, **kwargs): ...
        def foo(a, *args: int, **fooo: str,): ...
        def foo(a, /, b: i, c: i=3, **kwargs): ...
        def foo(*args): ...
        ");
    lambda_simple: dedent("
        lambda: ...
        lambda a: ...
        lambda a, b,: ...
        lambda a=3, b=4: ...
        ");
    lambda_slash: dedent("
        lambda a, /, b=3: ...
        lambda a=3, /,: ...
        lambda a=3, /: ...
        lambda a=3, /, *args, c=3: ...
        lambda a=3, /, c=3: ...
        lambda a, /, b, c=3, **kwargs: ...
        lambda a, a=3, /, c=3, *, d: ...
        ");
    lambda_args: dedent("
        lambda a, *, b=3: ...
        lambda a=3, *, c: ...
        lambda a, *, x, **kwargs: ...
        lambda a, *args, **fooo ,: ...
        lambda *args: ...
        ");
    lambda_kwargs: dedent("
        lambda **kwargs: ...
        lambda a=3, **kwargs: ...
        lambda a, **kwargs: ...
        lambda a, /, b, *, x, **kwargs: ...
        ");
    lambda_fails: dedent("
        lambda a, *, **kwargs: ...
        lambda a=3, b: ...
        lambda **a, b: ...
        lambda a=3, *,: ...
        lambda ,: ...
        ");
    lambda_fail_with_slash: dedent("
        lambda a=3, /, b: ...
        lambda a=3, /, b=3, c: ...
        lambda a, /, b=3, c: ...
        lambda z, a=3, d, /, b: ...
        ");
    with_items_simple: dedent("
        with foo as bar: ...
        with foo as bar, baz: ...
        ");
    with_items_parentheses: dedent("
        with (foo as bar): ...
        with (foo as bar, baz): ...
        with ((foo)): ...
        with (i for i in 42): ...
        with (yield y): ...
        ");
    string_literals_in_strings: dedent(r#"
        '"""'
        "'''"
        ""
        ''
        "#);
    bytes: dedent(r#"
        b'asdf\n\''
        br'\asdf'
        Br'asdf"""'
        bR''
        BR'\''
        "#);
    exception_group: dedent(r#"
        try:
            ...
        except* SpamError:
            ...
        except* FooError as e:
            ...
        except* (BarError, BazError) as e:
            ...
        "#);
    invalid_syntax: dedent(r#"
        def f():
            (x,) += 1
            z
        "#);
);
