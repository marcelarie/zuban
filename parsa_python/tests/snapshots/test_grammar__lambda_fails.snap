---
source: parsa_python/tests/test_grammar.rs
expression: tree_to_string(tree)
---
file: 0-100
 Error(stmt): 1-9
  Error(simple_stmts): 1-9
   Error(simple_stmt): 1-9
    Error(expr_stmt): 1-9
     Error(star_expressions): 1-9
      Error(expression): 1-9
       Error(lambda): 1-9
        Keyword: 1-7 "lambda"
        Error(lambda_parameters): 8-9
         Error(lambda_param_with_default): 8-9
          Name: 8-9 "a"
 Error(stmt): 9-10
  Error(Keyword): 9-10 ","
 Error(stmt): 11-12
  Error(simple_stmts): 11-12
   Error(simple_stmt): 11-12
    Error(expr_stmt): 11-12
     Error(star_expressions): 11-12
      Error(star_expression): 11-12
       Keyword: 11-12 "*"
 Error(stmt): 12-13
  Error(Keyword): 12-13 ","
 Error(stmt): 14-16
  Error(Keyword): 14-16 "**"
 stmt: 16-28
  simple_stmts: 16-28
   simple_stmt: 16-27
    expr_stmt: 16-27
     star_expressions: 16-22
      expression: 16-22
       bitwise_or: 16-22
        Name: 16-22 "kwargs"
     annassign: 22-27
      Keyword: 22-23 ":"
      expression: 24-27
       bitwise_or: 24-27
        Keyword: 24-27 "..."
   Newline: 27-28 "\n"
 Error(stmt): 28-41
  Error(simple_stmts): 28-41
   Error(simple_stmt): 28-41
    Error(expr_stmt): 28-41
     Error(star_expressions): 28-41
      Error(expression): 28-41
       Error(lambda): 28-41
        Keyword: 28-34 "lambda"
        Error(lambda_parameters): 35-41
         lambda_param_with_default: 35-38
          Name: 35-36 "a"
          Keyword: 36-37 "="
          expression: 37-38
           bitwise_or: 37-38
            Number: 37-38 "3"
         Keyword: 38-39 ","
         Error(lambda_param_with_default): 40-41
          Name: 40-41 "b"
 Error(stmt): 41-42
  Error(Keyword): 41-42 ":"
 stmt: 43-47
  simple_stmts: 43-47
   simple_stmt: 43-46
    expr_stmt: 43-46
     star_expressions: 43-46
      expression: 43-46
       bitwise_or: 43-46
        Keyword: 43-46 "..."
   Newline: 46-47 "\n"
 Error(stmt): 47-58
  Error(simple_stmts): 47-58
   Error(simple_stmt): 47-58
    Error(expr_stmt): 47-58
     Error(star_expressions): 47-58
      Error(expression): 47-58
       Error(lambda): 47-58
        Keyword: 47-53 "lambda"
        lambda_parameters: 54-58
         lambda_star_etc: 54-58
          lambda_double_starred_param: 54-57
           Keyword: 54-56 "**"
           Name: 56-57 "a"
          Keyword: 57-58 ","
 stmt: 59-66
  simple_stmts: 59-66
   simple_stmt: 59-65
    expr_stmt: 59-65
     star_expressions: 59-60
      expression: 59-60
       bitwise_or: 59-60
        Name: 59-60 "b"
     annassign: 60-65
      Keyword: 60-61 ":"
      expression: 62-65
       bitwise_or: 62-65
        Keyword: 62-65 "..."
   Newline: 65-66 "\n"
 Error(stmt): 66-80
  Error(simple_stmts): 66-80
   Error(simple_stmt): 66-80
    Error(expr_stmt): 66-80
     Error(star_expressions): 66-80
      Error(expression): 66-80
       Error(lambda): 66-80
        Keyword: 66-72 "lambda"
        Error(lambda_parameters): 73-80
         lambda_param_with_default: 73-76
          Name: 73-74 "a"
          Keyword: 74-75 "="
          expression: 75-76
           bitwise_or: 75-76
            Number: 75-76 "3"
         Keyword: 76-77 ","
         Error(lambda_star_etc): 78-80
          Keyword: 78-79 "*"
          Keyword: 79-80 ","
 Error(stmt): 80-81
  Error(Keyword): 80-81 ":"
 stmt: 82-86
  simple_stmts: 82-86
   simple_stmt: 82-85
    expr_stmt: 82-85
     star_expressions: 82-85
      expression: 82-85
       bitwise_or: 82-85
        Keyword: 82-85 "..."
   Newline: 85-86 "\n"
 Error(stmt): 86-92
  Error(simple_stmts): 86-92
   Error(simple_stmt): 86-92
    Error(expr_stmt): 86-92
     Error(star_expressions): 86-92
      Error(expression): 86-92
       Error(lambda): 86-92
        Keyword: 86-92 "lambda"
 Error(stmt): 93-94
  Error(Keyword): 93-94 ","
 Error(stmt): 94-95
  Error(Keyword): 94-95 ":"
 stmt: 96-100
  simple_stmts: 96-100
   simple_stmt: 96-99
    expr_stmt: 96-99
     star_expressions: 96-99
      expression: 96-99
       bitwise_or: 96-99
        Keyword: 96-99 "..."
   Newline: 99-100 "\n"
 Endmarker: 100-100 ""

